crates/compiler/builtins/roc/Hash.roc

```diff
@@ -73,12 +75,38 @@ Hasher has
     ## accumulated hash state.
     complete : a -> U64 | a has Hasher
 
+    ## Resets the internal state of a hasher
+    ## The hasher should still have all the same parameters and seeds.
+    reset : a -> a | a has Hasher
+
 ## Adds a string into a [Hasher] by hashing its UTF-8 bytes.
 hashStrBytes = \hasher, s ->
-    Str.walkUtf8WithIndex s hasher \accumHasher, byte, _ ->
-        addU8 accumHasher byte
+    addBytes hasher (Str.toUtf8 s)
 
 ## Adds a list of [Hash]able elements to a [Hasher] by hashing each element.
 hashList = \hasher, lst ->
     List.walk lst hasher \accumHasher, elem ->
         hash accumHasher elem
+
+## Adds a container of [Hash]able elements to a [Hasher] by hashing each element.
+## The container is iterated using the walk method passed in.
+## The order of the elements does not effect the final hash.
+hashUnordered = \hasher, container, walk ->
+    walk
+        container
+        0
+        (\accum, elem ->
+            x =
+                hasher
+                |> reset
+                |> hash elem
+                |> complete
+            nextAccum = Num.addWrap accum x
```

Comment:
Is there an upside to accumulating into a U128 here, to reduce collision space? Or is that a false analysis because we then have to push out into a U64 anyway?

Reply:
Technically there should be a minor advantage to accumulating in a U128 from the hash space perspective. Instead of merging the extra overflow bit entry back in via addition, it would be merged back in with proper mixing and hashing.

That said, I don't think it will make a tangible difference. Also, it would cost a tiny bit in performance. This implementation is what absl does, it is what I tend to default to when I am not sure what to pick. I at least know that is battle tested.

Gonna leave it be.